cpp/src/Ice/ObjectAdapterI.cpp

```diff
@@ -538,10 +527,20 @@ Ice::ObjectAdapterI::findServantLocator(const string& prefix) const
     return _servantManager->findServantLocator(prefix);
 }
 
-ObjectPtr
-Ice::ObjectAdapterI::dispatcher() const noexcept
+const ObjectPtr&
+Ice::ObjectAdapterI::dispatchPipeline() const noexcept
 {
-    return _servantManager;
+    lock_guard lock(_mutex);
+    if (!_dispatchPipeline)
+    {
+        _dispatchPipeline = _servantManager;
+        while (!_middlewareFactoryStack.empty())
+        {
+            _dispatchPipeline = _middlewareFactoryStack.top()(_dispatchPipeline);
```

Comment:
can we move `_dispatchPipeline` here?
```suggestion
            _dispatchPipeline = _middlewareFactoryStack.top()(std::move(_dispatchPipeline));
```